petab/v2/lint.py

```diff
@@ -0,0 +1,549 @@
+                logger.warning(result.message)
+            elif result.level >= ValidationIssueSeverity.ERROR:
+                logger.error(result.message)
+
+        if not self:
+            logger.info("PEtab format check completed successfully.")
+
+    def has_errors(self) -> bool:
+        """Check if there are any errors in the validation results."""
+        return any(
+            result.level >= ValidationIssueSeverity.ERROR for result in self
+        )
+
+
+def lint_problem(problem: Problem | str | Path) -> ValidationResultList:
+    """Validate a PEtab problem.
+
+    Arguments:
+        problem: PEtab problem to check. Instance of :class:`Problem` or path
+        to a PEtab problem yaml file.
+    Returns:
+        A list of validation results. Empty if no issues were found.
+    """
+
+    if isinstance(problem, str | Path):
+        problem = Problem.from_yaml(str(problem))
+
+    return problem.validate()
+
+
+class ValidationTask(ABC):
+    """A task to validate a PEtab problem."""
+
+    @abstractmethod
+    def run(self, problem: Problem) -> ValidationIssue | None:
+        """Run the validation task.
+
+        Arguments:
+            problem: PEtab problem to check.
+        Returns:
+            Validation results or ``None``
+        """
+        ...
+
+    def __call__(self, *args, **kwargs):
+        return self.run(*args, **kwargs)
+
+
+class CheckModel(ValidationTask):
+    """A task to validate the model of a PEtab problem."""
+
+    def run(self, problem: Problem) -> ValidationIssue | None:
+        if problem.model is None:
+            return ValidationError("Model is missing.")
+
+        if not problem.model.is_valid():
+            # TODO get actual model validation messages
+            return ValidationError("Model is invalid.")
+
+
+class CheckTableExists(ValidationTask):
+    """A task to check if a table exists in the PEtab problem."""
+
+    def __init__(self, table_name: str):
+        if table_name not in ["measurement", "observable", "parameter"]:
+            # all others are optional
+            raise ValueError(
+                f"Table name {table_name} is not supported. "
+                "Supported table names are 'measurement', 'observable', "
+                "'parameter'."
+            )
+        self.table_name = table_name
+
+    def run(self, problem: Problem) -> ValidationIssue | None:
+        if getattr(problem, f"{self.table_name}_df") is None:
+            return ValidationError(f"{self.table_name} table is missing.")
+
+
+class CheckMeasurementTable(ValidationTask):
+    """A task to validate the measurement table of a PEtab problem."""
+
+    def run(self, problem: Problem) -> ValidationIssue | None:
+        if problem.measurement_df is None:
+            return
+
+        try:
+            check_measurement_df(problem.measurement_df, problem.observable_df)
+
+            if problem.condition_df is not None:
+                # TODO: handle missing condition_df
```

Comment:
:+1: fine for now but I think some `DEBUG` or `INFO` message would be good, so a user knows not all aspects of the table were checked.

Reply:
This was about handling the (iirc) optional conditions table in petab2 (which means, all conditionIds will be empty and model values are to be used). This will require various changes. I guess I will open a separate issue on that.
So in this case it's not about something going unnoticed, but rather a false positive.

Reply:
Ah, then my comment is for the `if problem.*_df is None: return` in other validators.

Reply:
I'd say those are fine, because either the tables are required and `CheckTableExists` will emit an error, or those are optional. I am not sure if we want messages for missing optional tables, but I don't have a strong opinion there. If they aren't there but should be there, I'd say this will most likely cause downstream errors.
For validation of v1 problems we have a warning on missing visualization tables, but I regret having added that :sweat_smile:.

Reply:
Could be better to implement this like so then
```python
class CheckConditionTable(ValidationTask):
    requires=[CheckTableExists("condition")]
    ...
```

and remove `CheckTableExists("condition")` from `default_validation_tasks`. This would implement the dependency graph though -- fine to keep out of this PR

Reply:
Yes, a dependency graph would be convenient to run as many checks as possible in a single run. However, I think maintaining that properly will be tedious. For now, I'd leave it as is.